libsplinter/src/admin/store/diesel/operations/add_circuit.rs

```diff
@@ -0,0 +1,200 @@
+// Copyright 2018-2020 Cargill Incorporated
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Provides the "add circuit" operation for the `DieselAdminServiceStore`.
+
+use diesel::{dsl::insert_into, prelude::*, result::Error::NotFound};
+
+use super::AdminServiceStoreOperations;
+use crate::admin::store::{
+    diesel::{
+        models::{
+            CircuitMemberModel, CircuitModel, ServiceAllowedNodeModel, ServiceArgumentModel,
+            ServiceModel,
+        },
+        schema::{circuit, circuit_member, service, service_allowed_node, service_argument},
+    },
+    error::AdminServiceStoreError,
+    Circuit, CircuitNode,
+};
+
+pub(in crate::admin::store::diesel) trait AdminServiceStoreAddCircuitOperation {
+    fn add_circuit(
+        &self,
+        circuit: Circuit,
+        nodes: Vec<CircuitNode>,
+    ) -> Result<(), AdminServiceStoreError>;
+}
+
+#[cfg(feature = "postgres")]
+impl<'a> AdminServiceStoreAddCircuitOperation
+    for AdminServiceStoreOperations<'a, diesel::pg::PgConnection>
+{
+    fn add_circuit(
+        &self,
+        circuit: Circuit,
+        nodes: Vec<CircuitNode>,
+    ) -> Result<(), AdminServiceStoreError> {
+        if circuit::table
+            .filter(circuit::circuit_id.eq(&circuit.id))
+            .first::<CircuitModel>(self.conn)
+            .map(Some)
+            .or_else(|err| if err == NotFound { Ok(None) } else { Err(err) })
+            .map_err(|err| AdminServiceStoreError::QueryError {
+                context: String::from("Error occurred fetching Circuit"),
+                source: Box::new(err),
+            })?
+            .is_some()
```

Comment:
Instead of `if option.is_some() { ... }`, you can do `option.ok_or_else(|| { ... })`

Reply:
The point of the `if option.is_some()` is to check if the query returns a circuit (`is_some()`), in which case we need to return an error (because we're trying to add a circuit, but one already exists). So, it wouldn't make sense, in this scenario, to use `option.ok_or_else`.

Reply:
Ah, I misread the code. 👍